src/Microsoft.AspNetCore.Sockets/DefaultConnectionContext.cs

```diff
@@ -75,14 +75,20 @@ public DefaultConnectionContext(string id, Channel<byte[]> transport, Channel<by
 
         public void OnHeartbeat(Action<object> action, object state)
         {
-            _heartbeatHandlers.Add((action, state));
```

Comment:
Why isn't this a concurrent dictionary?

Reply:
It could be a `ConcurrentBag`. Or, as I suggested originally, an `ImmutableList` :trollface: 

It's rarely updated though. For the most part, handlers are added at the start and then run, so the lock is pretty low impact. Concurrent collections have their own trade-offs but tbh I'm not sure which is more appropriate now.